src/IssueViz.Security/OpenInIDE/Api/StatusRequestHandler.cs

```diff
@@ -31,22 +31,34 @@ namespace SonarLint.VisualStudio.IssueVisualization.Security.OpenInIDE.Api
     [Export(typeof(IStatusRequestHandler))]
     internal class StatusRequestHandler : IStatusRequestHandler
     {
-        private readonly IVsShell vsShell;
-        private readonly IVsSolution vsSolution;
+        private readonly IVsUIServiceOperation vSServiceOperation;
 
         [ImportingConstructor]
-        public StatusRequestHandler([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        public StatusRequestHandler(IVsUIServiceOperation vSServiceOperation)
         {
-            vsShell = serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
-            vsSolution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            this.vSServiceOperation = vSServiceOperation;
         }
 
         Task<IStatusResponse> IStatusRequestHandler.GetStatusAsync()
         {
-            vsShell.GetProperty((int) __VSSPROPID5.VSSPROPID_AppBrandName, out var ideName);
+            var result = vSServiceOperation.Execute<SVsShell, IVsShell, Task<IStatusResponse>>(
```

Comment:
I mean, this is not the ugliest thing ever considering theres only 2 services we need however I am unsure how I'd l^solve this if there were more than 2

Reply:
It's a bit surprising we haven't hit an example that uses multiple services before...

One option would be to change the `IVsUIServiceOperation` interface (or create a separate interface) that is designed to work with multiple services
e.g. by passing `IServiceProvider` into the callback instead of a specific service, so the consuming could request any services it likes. The service operation class would be responsible for handling the thread transitions, and the consuming code would still be simpler as it would still only import the `IVsUIServiceOperation` interface.

As it is, the current code could be structured differently to make it more readable by avoiding the nesting e.g.

```
        async Task<IStatusResponse> IStatusRequestHandler.GetStatusAsync()
        {
            var ideName = await GetIdeNameAsync();
            var ideInstance = await GetIdeInstanceAsync();

            return new StatusResponse(ideName, ideInstance);
        }

        private async Task<string> GetIdeNameAsync()
            => await vSServiceOperation.ExecuteAsync<SVsShell, IVsShell, string>(
                shell =>
                {
                    shell.GetProperty((int)__VSSPROPID5.VSSPROPID_AppBrandName, out var ideName);
                    return ideName as string ?? "Microsoft Visual Studio";
                });

        private async Task<string> GetIdeInstanceAsync()
            => await vSServiceOperation.ExecuteAsync<SVsSolution, IVsSolution, string>(
                solution =>
                {
                    solution.GetProperty((int)__VSPROPID.VSPROPID_SolutionBaseName, out var ideInstance);
                    return ideInstance as string ?? "";
                });
```

This version is slightly less efficient as it will have to acquire the UI thread twice, but it won't be noticeable in practice.